File: code/utils/data_paths.R
# Path helpers backed by data/registry.json. From any script:
#   source("code/utils/data_paths.R")
#   df <- arrow::read_parquet(data_path("crsp_msf_2000_2023"))
#   arrow::write_parquet(df, out_path("panel_clean", stage = "processed", depends_on = "crsp_msf_2000_2023"))
# Never write literal Dropbox / home-directory paths in analysis code.

.cloco_root <- function() {
  d <- getwd()
  for (i in 1:6) {
    if (file.exists(file.path(d, "data", "registry.json"))) return(d)
    d <- dirname(d)
  }
  stop("data/registry.json not found above the working directory")
}

.cloco_env <- function() {
  root <- .cloco_root()
  env <- list(PROJECT_ROOT = root)
  f <- file.path(root, ".env")
  if (file.exists(f)) {
    for (ln in readLines(f, warn = FALSE)) {
      ln <- trimws(ln)
      if (ln == "" || startsWith(ln, "#") || !grepl("=", ln)) next
      kv <- strsplit(ln, "=", fixed = TRUE)[[1]]
      val <- trimws(paste(kv[-1], collapse = "="))
      if (val != "") env[[trimws(kv[1])]] <- path.expand(gsub("^['\"]|['\"]$", "", val))
    }
  }
  for (v in c("DROPBOX_ROOT", "DATA_ROOT")) {
    sysv <- Sys.getenv(v, unset = "")
    if (sysv != "") env[[v]] <- path.expand(sysv)
  }
  if (is.null(env$DROPBOX_ROOT)) {
    for (g in c("~/Dropbox", "~/Library/CloudStorage/Dropbox")) if (dir.exists(path.expand(g))) { env$DROPBOX_ROOT <- path.expand(g); break }
  }
  env
}

.cloco_resolve <- function(template) {
  env <- .cloco_env()
  vars <- regmatches(template, gregexpr("\\$\\{[A-Z0-9_]+\\}", template))[[1]]
  for (v in unique(vars)) {
    key <- gsub("[${}]", "", v)
    if (is.null(env[[key]])) stop(sprintf("%s not set in .env (see .env.example)", v))
    template <- gsub(v, env[[key]], template, fixed = TRUE)
  }
  if (!grepl("^(/|[A-Za-z]:)", template)) template <- file.path(env$PROJECT_ROOT, template)
  normalizePath(template, mustWork = FALSE)
}

data_path <- function(name, must_exist = TRUE) {
  if (requireNamespace("jsonlite", quietly = TRUE)) {
    reg <- jsonlite::fromJSON(file.path(.cloco_root(), "data", "registry.json"), simplifyVector = FALSE)
    if (is.null(reg$datasets[[name]])) stop(sprintf("'%s' is not in data/registry.json — register it with /data-registry add", name))
    p <- .cloco_resolve(reg$datasets[[name]]$path)
  } else {  # no jsonlite: ask the registry CLI to resolve the path
    p <- suppressWarnings(system2("python3", shQuote(c(file.path(.cloco_root(), ".claude/scripts/data_registry.py"), "where", name)), stdout = TRUE, stderr = FALSE))
    if (length(p) == 0 || !nzchar(p[1])) stop(sprintf("'%s' is not in data/registry.json (or a ${ROOT} is unset in .env)", name))
    p <- p[1]
  }
  if (must_exist && !file.exists(p)) stop(sprintf("%s is registered at %s but the file is not on this machine", name, p))
  p
}

out_path <- function(name, stage = "processed", fmt = "parquet", root = "${PROJECT_ROOT}",
                     depends_on = NULL, source = NULL, register = TRUE) {
  template <- if (root == "${PROJECT_ROOT}") sprintf("%s/data/%s/%s.%s", root, stage, name, fmt) else sprintf("%s/%s.%s", root, name, fmt)
  p <- .cloco_resolve(template)
  dir.create(dirname(p), recursive = TRUE, showWarnings = FALSE)
  if (register) {
    script <- tryCatch(basename(sys.frame(1)$ofile), error = function(e) "R session")
    args <- c(file.path(.cloco_root(), ".claude/scripts/data_registry.py"), "add", name, "--path", template,
              "--stage", stage, "--source", if (is.null(source)) paste("produced by", script) else source,
              "--format", fmt, "--force")
    if (!is.null(depends_on)) args <- c(args, "--depends", depends_on)
    system2("python3", shQuote(args), stdout = FALSE, stderr = FALSE)
  }
  p
}

File: code/utils/data_paths.py
"""
Path helpers backed by data/registry.json. Import from any script under code/:

    import sys; sys.path.insert(0, "code")          # if not running from repo root
    from utils.data_paths import data_path, out_path, project_root

    df = pd.read_parquet(data_path("crsp_msf_2000_2023"))
    df.to_parquet(out_path("panel_clean", stage="processed", depends_on=["crsp_msf_2000_2023"]))

Never write literal Dropbox / home-directory paths in analysis code.
"""
from __future__ import annotations

import subprocess
import sys
from pathlib import Path

_ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(_ROOT / ".claude" / "scripts"))
from data_registry import load_env, load_registry, resolve  # noqa: E402


def project_root() -> Path:
    return _ROOT


def data_path(name: str, must_exist: bool = True) -> Path:
    """Absolute path of a registered dataset on this machine."""
    reg = load_registry()["datasets"]
    if name not in reg:
        raise KeyError(f"'{name}' is not in data/registry.json — register it with /data-registry add")
    p, missing = resolve(reg[name]["path"], load_env())
    if missing:
        raise EnvironmentError(f"{missing} not set in .env (see .env.example)")
    if must_exist and not p.exists():
        raise FileNotFoundError(f"{name} is registered at {p} but the file is not on this machine")
    return p


def out_path(name: str, stage: str = "processed", fmt: str = "parquet", root: str = "${PROJECT_ROOT}",
             depends_on: list[str] | None = None, source: str | None = None, register: bool = True) -> Path:
    """Path to write a derived dataset to; registers it (idempotently) so downstream code can data_path() it."""
    template = f"{root}/data/{stage}/{name}.{fmt}" if root == "${PROJECT_ROOT}" else f"{root}/{name}.{fmt}"
    p, missing = resolve(template, load_env())
    if missing:
        raise EnvironmentError(f"{missing} not set in .env")
    p.parent.mkdir(parents=True, exist_ok=True)
    if register:
        cmd = [sys.executable, str(_ROOT / ".claude/scripts/data_registry.py"), "add", name, "--path", template,
               "--stage", stage, "--source", source or f"produced by {Path(sys.argv[0]).name}", "--format", fmt, "--force",
               "--script", str(Path(sys.argv[0]).resolve().relative_to(_ROOT)) if Path(sys.argv[0]).resolve().is_relative_to(_ROOT) else Path(sys.argv[0]).name]
        if depends_on:
            cmd += ["--depends", *depends_on]
        subprocess.run(cmd, check=False, capture_output=True)
    return p
